### DISCUSSION TITLE
HID Joystick mode - Implementation and design questions

### DISCUSSION THREAD

**Discussion by countofkrakow** [ExpressLRS/ExpressLRS#2168] [Ideas]
Hi guys,

I have been trying to build a usb joystick dongle for ELRS with the project [here](https://github.com/mikeneiderhauser/CRSFJoystick/tree/master). I got a working simulator dongle but got stuck on flashing the rx through usb so I asked Jye Smith for help and he informed me that I could implement USB HID an an output mode on ELRS receivers based on [this example](https://github.com/ExpressLRS/ExpressLRS/pull/2140/files#diff-2b446cb0067d010f86af3114c480af171cff5fdad35e88ed7cde3d30651bb11a) so I could make a simulator dongle by just connecting an USB HID Serial adapter to a reciever.

So I went back to CRSFJoystick which uses picoGamepad and looked at how HID communication is done there and boy do I have questions:

1. How important are HID descriptors and are they even relevant here? I have noticed that when I plug in a crsfJoystick device that none of the descriptor information from [here](https://gitlab.com/realrobots/PicoGamepad/-/blob/master/PicoGamepad.cpp) actually makes it to my pc and it still reads as a raspberry pi pico. It appears as a com port device rather than a usb gamepad. Works just fine for simulator use; however, maybe this doesn't matters since the sim just asks you to assign channels? Another thing here is whether there is a spot in the ELRS code to return an HID descriptor where prompted. I don't see crsfjoystick using the code for PicoGamepad's HID descriptor anywhere but I may be misreading the code so let me know. 
2. On pg 24/25, the USB standard [doc ](https://www.usb.org/sites/default/files/documents/hut1_12v2.pdf) describes the general message format of an HID report. The buffer starts with a an unsigned byte detailing the length of the vendor defined section, and then [0x01 for the second byte?](https://gitlab.com/realrobots/PicoGamepad/-/blame/master/PicoGamepad.cpp#L334) After that is vendor defined data. My question is: if the message length byte is only supposed to be a byte, then why is the length a uint32_t on line 89 in the [arduino mbed library](https://github.com/arduino/ArduinoCore-nRF528x-mbedos/blob/master/libraries/USBHID/USBHID_Types.h), which is used by PicoGamepad for the usb buffer? Also does the second byte have any importance? Can I just keep it at 0x01?

3. CRSFJoystick uses the full set of fields available offered by PicoGamepad. This includes hat buttons, which I have never encountered a use for in fpv controls also 8 full resolution channels, 4 of which are unused. I feel that, for elrs purposes, this could be pared down to 4 axis channels and 8 button channels. Is there any reason not to do this if we're not adhering to a standard gamepad format? This other HID controller library seems to implement channels as arbitrary and [optional](https://github.com/MHeironimus/ArduinoJoystickLibrary/blob/12cf2bbdb8910619d32ba3bf4b7d669c8e813b99/src/Joystick.cpp#L643).

help and information appreciated!

**Comment by AlessandroAU** [comment]
Hi, are you talking about having the ELRS receiver talk USB naively through it's output pins? 

The main issue I can see is that the Pi Pico has dedicated hardware and some nice abstraction layers to talk USB, the ESP8266 used in receivers does not. 

USB signalling is also a bit different to standard UART. To begin with it's differential and USB 2.0 uses ±400-mV voltage levels. **In theory**, I guess you could bit-bang USB 1.1 since that uses 3.3v. I found a project that does it, but it looks like absolute hell to integrate into ELRS and I'm sure you'd need to be an expert debugger to sort out all the possible ISR issues and race conditions. https://github.com/cnlohr/espusb Not mention write and debug your own HID stack. 

Unfortunately, ESP32 doesn't have a USB Peripheral either. It would likely be easier to port ELRS to the Pi Pico than add USB to a native ELRS ESP8266/ESP32 receiver. However, some of the ARM CPU based ELRS RX targets do have USB support, so it might be possible to do something with those?, I highly doubt the correct pins are broken out so you'd still need a custom board most likely.

**Reply by countofkrakow** [reply]
I was planning on sticking a USB HID to serial dongle between the rx and a pc. Thinking something based on the [cpl2102 chipset](https://www.silabs.com/interface/usb-bridges/classic/device.cp2102?tab=specs) since that's what the Betafpv [dongle ](https://betafpv.com/products/expresslrs-recovery-dongle)uses. Do you think that would solve the issue?

**Reply by AlessandroAU** [reply]
Well, does the CP2102 support a HID mode? I think it's just a serial bridge.

**Reply by countofkrakow** [reply]
Looked again and I think you are correct on it not having USB HID to Uart functionality. This [one ](https://www.amazon.com/MakerSpot-CP2110-Serial-Adapter-Dongle/dp/B01IVJEH82?th=1)seems to have it though.

**Reply by AlessandroAU** [reply]
Yes and no, it won't do what you want. Honestly, the Pico is a already a good choice. 
![image](https://user-images.githubusercontent.com/22009829/232379945-e334928e-7110-4a24-bda2-346ba8722b20.png)

**Reply by countofkrakow** [reply]
In that case you're suggesting the easiest path forward is to just port ELRS onto the Pico. Am I understanding that correctly?

My original plan was to simply get passthrough flash to work on CRSFJoystick. I opened an [issue ](https://github.com/CapnBry/CRServoF/issues/29)on CRServoJoystick since CapnBry had the gall to make a serial passthru function that doesn't work when ported to Arduino lol.

Could you take a look and give your opinion on the passthru flashing difficulty I'm having?

Thanks and I genuinely appreciate the help

**Comment by L1cardo** [comment]
Are you still working on this?

**Reply by countofkrakow** [reply]
> Are you still working on this?

It's in a stable state on my branch for the squid stick.

**Reply by L1cardo** [reply]
I am thinking since now ELRS supports ESP32-S3 now, and S3 has a USB port, can we remove rp2040, just use S3 to output Joystick command?

**Reply by countofkrakow** [reply]
> I am thinking since now ELRS supports ESP32-S3 now, and S3 has a USB port, can we remove rp2040, just use S3 to output Joystick command?

It would require merging this project with ExpressLRS. If you think that's feasible, feel free to submit a PR.

**Reply by L1cardo** [reply]
> > I am thinking since now ELRS supports ESP32-S3 now, and S3 has a USB port, can we remove rp2040, just use S3 to output Joystick command?
> 
> It would require merging this project with ExpressLRS. If you think that's feasible, feel free to submit a PR.

They would be beyond my capabilities 😂
I know little about coding
